[1단계 - 페이먼츠] - 데이지(김유경) 미션 제출합니다.

kimyou1102 opened the pull request:
안녕하세요 쵸파! 레벨1에 이어서 이번 미션에도 또 만났네요😊
PR 본문에 쵸파의 의견이 궁금한 부분들은 작성하였습니다!
이번 미션도 잘 부탁드립니다🙇‍♀️

## 🎯 미션 소개

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

- 재사용 가능한 Input Component를 개발한다.
- Storybook을 사용하여 컴포넌트의 다양한 상태를 시각적으로 테스트한다.
- 카드 정보를 효과적으로 렌더링 하기 위한 상태 관리를 경험한다.

---

## 🕵️ 셀프 리뷰(Self-Review)

### 제출 전 체크 리스트

- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] 기본적인 프로그래밍 요구 사항(코드 컨벤션, 에러 핸들링 등)을 준수했나요?
- [x] 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 기본 배포 링크 기입:[배포링크](https://kimyou1102.github.io/react-payments/)
  - 스토리북 배포 링크 기입: [배포링크](https://6800afab6301b41dd9299995-ngvjwaglah.chromatic.com/)

![image](https://github.com/user-attachments/assets/41767a9c-74ac-4da2-adc8-ef9f1b1a1d3f)

### 리뷰 요청 & 논의하고 싶은 내용

#### 1) 이번 단계에서 가장 많이 고민했던 문제
카드 정보에 대한 input의 타입을 2가지중 고민했습니다.

1. tyep=”text”
2. type=”number”

타입을 text로 하면 사용자의 입력에 대한 자유도가 높아서 숫자로 제한하는 것이 맞다고 생각됐습니다. 그래서 number타입을 선택했습니다. 그리고 input의 value의 state의 타입도 number로 두었습니다. 그런데 
이렇게 하면 앞자리가 `0`인 값(예: `0123`)이 `123`으로 변환되어 `0`이 사라지는 문제가 있었습니다.

그래서 이를 해결하기 위해 `input type="number"`는 유지하되, input태그의 value의 state는 `string`으로 변경하였습니다. 그래서 `value`에 문자열 숫자(예: `"0123"`)를 그대로 저장되도록 했습니다.

이렇게 하면 0으로 시작하는 숫자에 대한 인식도 되면서 사용자 입력의 제한도 원하는 대로 됩니다. input의 value는 문자열이니 이렇게 해도 맞다고 생각이 들어서 이런 방식을 선택했는데 사용자 입력을 제한 하는 방식에 대한 쵸파의 의견이 궁금합니다!

#### 2) 이번 리뷰를 통해 논의하고 싶은 부분
도메인과 UI를 분리하기 위해 비즈니스 로직을 커스텀훅으로 분리했습니다. 이렇게 분리하는 과정에서 훅을 하나로 할지, 각 입력값에 대한 훅을 만들지 고민했습니다. 

1. 훅을 하나로 했을때 
    - 장점: 카드정보에 대한 값이 하나의 파일에서 관리되기에 응집성이 높아진다.
    - 단점: 훅에서 리턴할 값들이 너무 많아지고 유지보수성이 떨어질 수 있다(훅 내부의 복잡성 증가)
2. 훅을 여러개로 쪼갰을때(카드번호, 카드유효기간, 카드 CVC)
    - 장점: 훅의 복잡성이 절감되어 유지보수성이 높아진다.
    - 단점: 여러곳에서 관리하게 되므로 응집성이 떨어질 수 있다.

고민끝에 위와 같이 정리를 했고 여러개로 쪼개는 방식을 선택했습니다.
이렇게 나눈 것에 대한 쵸파의 의견과 쵸파가 역할과 책임을 나누는 기준이 궁금합니다!

#### 3) 리팩토링 시 할 목록
- 입력된 유효기간의 년도가 이번년도일때, 월 유효기간 입력을 이번달 이상으로 제한하는 것을 할 예정입니다.
- 상수화 파일 분리할 예정입니다.

---

## ✅ 리뷰어 체크 포인트

<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.
코드의 완성도만이 아니라, 리뷰이가 구현 과정에서 어떤 고민과 결정을 하며 학습했는지도 함께 고려해 주세요. -->

### 1. 컴포넌트 설계

- **재사용성**과 **확장성**을 고려해 Input 컴포넌트를 잘 설계했는가?
- 컴포넌트를 분리할 기준(역할, 책임, 관심사 등)을 고민한 흔적이 보이는가?
- UI 단위가 지나치게 작거나 크지 않고 적절한 수준으로 구성되어 있는가?

### 2. 상태 관리

- 카드 정보 렌더링에 필요한 상태를 React의 상태로 적절하게 관리하고 있는가?

### 3. Storybook 활용

- 컴포넌트의 다양한 상태(정상 입력, 오류, 빈값 등)를 확인할 수 있는 스토리가 작성되어 있는가?

### 4. UI/UX

- 유효성 검증, 에러 메시지 등에서 사용자 경험을 고려한 처리가 이루어졌는가?

dlsxjzld committed:
chore: ESLint 와 Prettier 적용

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
docs: 구현 기능 목록 작성

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
chore: emotion 설치

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
docs: README.md에 컴포넌트 구조 및 상태 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
feat: 공통 Input 컴포넌트 추가 및 스타일링

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
feat: Input 컴포넌트 스토리 생성

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
feat: 카드 번호 입력을 위한 CardNumber 컴포넌트 추가

dlsxjzld committed:
feat: CardNumber 스토리 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
style: Input 너비 속성 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
feat: CardNumberField 컴포넌트 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
feat: CardNumberField 스토리 생성

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
refactor: CardNumberField로 카드 번호 입력 컴포넌트 변경

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
feat: 카드 유효기간 입력을 위한 CardValidityPeriod 컴포넌트 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
refactor: 일관성을 위해 카드 유효기간 컴포넌트 이름에 Field 접미사 추가
- CardValidityPeriod에서 CardValidityPeriodField로 변경

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
feat: CardValidityPeriodField 스토리 생성

dlsxjzld committed:
feat: CardCVCField 컴포넌트 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
refactor: 카드 정보 입력의 description 옵셔널로 변경

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
feat: CardCVCField 스토리 생성

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
refactor: 명확한 네이밍을 위해 CardNumber에서 CardInputSection으로 이름 변경

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
refactor: 명확한 네이밍을 위해 cardInfo에서 cardInfoForm으로 이름 변경

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
refactor: 카드 관련 필드의 value 타입을 string으로 변경

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
feat: CardPreview 컴포넌트 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
refactor: CardCVC 기본값 빈 문자열로 변경 및 유효성 검사 조건 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
feat: CardPreview 스토리 생성

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
feat: 카드 입력 필드에 플레이스홀더 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
refactor: CVC 유효성 검사 로직 개선 및 min/max 속성 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
refactor: 카드 번호 min/max 속성 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
refactor: 카드 유효기간 min/max 속성 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
refactor: 카드 CVC 글자수 최대 3자리 제한 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
refactor: CardCVCField의 상태 위로 올리기

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
refactor: 카드 유효기간 스토리의 month/year 값을 문자열로 변경

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
refactor: CardCVCField 스토리에 상태 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
refactor: CardCVCField 스토리에 에러 스토리 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
refactor: 카드 입력 필드 스토리의 상태 문자열로 변경

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
refactor: CardInputSection 스토리에서 기본 스토리 변경 및 카드 번호 스토리 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
refactor: CardInputSection 스토리에서 카드 유효기간 스토리 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
refactor: CardInputSection 스토리에서 카드 CVC 스토리 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
refactor: CardCVCField 스토리에서 글자수 최대 3자리 제한 로직 변경

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
feat: App.tsx에 카드 입력 필드와 관련된 컴포넌트 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
feat: reset.css 파일 추가 및 기본 스타일 리셋 적용

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
style: input type number의 스핀 버튼 제거

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
remove: 안쓰는 파일 제거

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
style: 카드 미리보기 가운데 정렬

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
style: reset.css 파일로 스타일 초기화

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
style: main과 form 태그 스타일 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
fix: 언어 설정을 한국어로 변경

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
refactor: useCardNumber 커스텀 훅으로 분리

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
refactor: useCardValidityPeriod 커스텀 훅으로 분리

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
refactor: useCardCVC 커스텀 훅으로 분리

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
refactor: 카드 유효기간 검증 로직 간소화

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
feat: 카드 정보 입력 시 -+.eE 제한

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
refactor: 불필요한 주석 제거

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
refactor: 카드 번호 유효성 검사 로직 개선 및 오류 메시지 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
feat: CVC 오류 메시지 상태 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

dlsxjzld committed:
feat: 카드 유효기간 오류 메시지 상태 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

kimyou1102 committed:
refactor: 배포 이미지 경로 위해 수정

kimyou1102 committed:
chore: 배포 세팅

kimyou1102 committed:
docs: 구현완료 사항 체크표시

kimyou1102 committed:
chore: 스토리북 배포 세팅

bassyu commented on src/hooks/useCardValidityPeriod.tsx:
```diff
@@ -0,0 +1,71 @@
+    const validatePeriod = {
+      month: (value: string) => {
+        const isInvalidMonth =
+          Number.parseInt(value, 10) > 12 ||
```
Number 대신 parseInt를 활용해주신 이유가 궁금합니다~!

kimyou1102 replied:
input이 숫자만 입력가능해서 parseInt로 변환하든 Number로 하든 상관없다 생각해서 parseInt로 변환하게 됐어요. 그런데 의미적으로 parseInt를 하면 문자열 파싱의 의미를 드러낼 수 있어서 오히려 이경우는 Number로 변환하는 게 더 좋을 수 있겠네요!

bassyu reviewed (changes requested):
안녕하세요 데이지!
다시 만나게 되어서 반가워요~! 2레벨 첫 미션도 잘 부탁드려요!!

## 기능

이쁘게 잘 구현해주셨네요! 👍 

아래는 보완되면 좋은 것들 입니다!

1. 카드 번호를 입력할 때, 4의 배수일 때에는 에러 메세지가 나오지 않는게 어색하네요! 한글을 입력했는데, 숫자의 개수를 잘못 입력했다는 에러 메세지가 나옵니다!
2. <img width="100" alt="스크린샷 2025-04-18 오후 11 45 04" src="https://github.com/user-attachments/assets/65861d0f-7a76-4a66-ac1f-70e4ca39f55f" /> blur 후, 입력에 문제가 없다면 에러가 보이지 않으면 좋겠습니다~!

## 고민

> 사용자 입력을 제한 하는 방식

꼼꼼하게 고민 해주셨네요!  저도 문자열과 정규 표현식을 통해 0~9만 입력받을 수 있도록 제한할 것 같습니다!


> 이렇게 나눈 것에 대한 쵸파의 의견과 쵸파가 역할과 책임을 나누는 기준이 궁금합니다!

분리 기준을 잘 고민해주셨네요! 👍 
제가 해본다면.. 우선 각각의 입력마다 value와 error를 관리할 것 같습니다.
근데 이러면 useState를 너무 많이 사용하니, 객체로 바꾸고 useReducer를 활용할 것 같아요.

카드번호처럼 여러 입력 그룹이 의미를 가진다면, 각 입력을 한번 더 그룹화한 훅을 만들 것 같아요.

---

이번 미션도 수고 많으셨습니다! UX 우선적으로 개선해보면 좋을 것 같습니다!
리뷰 확인 후 다시 요청주세요~!

kimyou1102 commented on src/App.tsx:
```diff
@@ -1,11 +1,109 @@
+    errorMessage: cardValidityPeriodErrorMessage,
+  } = useCardValidityPeriod();
+
+  const {
+    cardCVC,
+    isCardCVCError,
+    onChangeCVC,
+    checkCardCVCError,
+    errorMessage: cardCVCErrorMessage,
+  } = useCardCVC();
```
원래는 cvc컴포넌트에서만 cvc정보가 필요하기에 상태를 컴포넌트 내부에 뒀습니다. 하지만 `CardInputSection` 컴포넌트의 errorMessage props를 전달할 때 error를 체크해주는 함수인 `checkCardCVCError` 가 필요해서 App으로 상태를 올렸습니다!

App에서 몰라도 되는 정보를 알게되면 정보들이 많아져서 복잡성이 증가하기 때문에 유지보수하기 어려워질 거 같아요!

이런 점을 피하기 위해서 상태를 필요한 곳에서만 좁게 사용해야 할 거 같습니다. 지금 코드에서 상태를 더 좁게 사용하기 위해서 `CardInputSection` 과 `CardCVCField` 를 묶은 새로운 컴포넌트를 만들고, 그 컴포넌트 내부에서만 cvc상태를 두는 것도 하나의 방법일 거 같네요!

bassyu replied:
넵 맞습니다! 복잡성은 중요한 포인트라고 생각해요 👍 리팩터링 방향성도 잘 잡으셨네요.

추가로, `cardCVC` 상태가 바뀌면 `App`컴포넌트랑 `App`의 자식 컴포넌트가 리렌더링되는 단점도 해결되겠네요~!

kimyou1102 commented on src/components/common/Input/Input.stories.tsx:
```diff
@@ -0,0 +1,52 @@
+    isError: false,
+    type: 'number',
+    placeholder: '카드번호를 입력하세요',
+    value: 0,
```
타입을 변경할때 미처 못했던 부분이 있네요😱 수정하겠습니다!

kimyou1102 replied:
수정커밋: [fix: 타입 에러 수정](https://github.com/woowacourse/react-payments/pull/439/commits/dfcbcfa107e206da9526206101ff211109ca2150)

kimyou1102 commented on src/components/cardInfoForm/CardValidityPeriodField/CardValidityPeriodField.tsx:
```diff
@@ -0,0 +1,75 @@
+          isError={isErrorMonth}
+          type="number"
+          name="cardValidityPeriod"
+          id={`cardValidityPeriod-${month}`}
```
컴포넌트가 여러번 사용될때 id가 중복될 수 있다는 것을 놓쳤네요
`CardValidityPeriodField`에서 useId() 훅을 사용해서 고유한 id를 만드는 방법으로 해결 가능할 거 같습니다!

kimyou1102 replied:
수정커밋: [refactor: useId훅을 이용해서 고유한 id값이 되도록 수정](https://github.com/woowacourse/react-payments/pull/439/commits/ed31837cc366ec5c532d9650c7043cace44dfcb7)

kimyou1102 commented on src/components/common/Input/Input.tsx:
props로 받은 정규표현식을 Input컴포넌트의 onChange에서 사용하면서 컴포넌트단에서 패턴 검증을 할 수 있도록 하는 것을 의미하는 걸까요??

```
function Input(props: InputProps) {
  const { onChange, ...rest } = props;
  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (new RegExp(props.regex).test(e.target.value)) {
      onChange(e);
    }
  };
  
  return <InputWrapper {...rest} onChange={handleChange} />;
}
```

bassyu replied:
넵 맞습니다! 👍 이건 예시 중 하나였고, 데이지가 직접 확장성을 고려한 기능을 추가해봐도 좋겠다는 뜻이었어요~!

kimyou1102 committed:
refactor: 카드 유효기간 입력 검증 리팩토링

- 현재보다 이전값을 유효기간으로 입력했을 시 예외처리 되도록 검증 추가
- 경우마다 다른 에러 메시지가 출력되도록 에러메시지 세분화

kimyou1102 committed:
fix: 타입 에러 수정

kimyou1102 commented on src/components/cardInfoForm/CardValidityPeriodField/CardValidityPeriodField.stories.tsx:
```diff
@@ -0,0 +1,146 @@
+    onChange: () => {},
+  },
+  render: (args) => {
+    const [cardValidityPeriod, setCardValidityPeriod] = useState({
```
커스텀훅을 사용하면서 초기값을 원하는 값으로 하기 위해서 args로 컴포넌트의 props값을 넘겨줬습니다. 이렇게 하니 input값이 변경되지 않았았어요.

초기값을 원하는 값으로 하기 위해 커스텀훅에서 초기값을 매개변수로 받게끔 변경하면 가능한데 스토리북을 위해 로직을 바꾸는 게 맞는지 고민이 됩니다!

kimyou1102 replied:
수정커밋: [refactor: 스토리북에서 커스텀훅 적용](https://github.com/woowacourse/react-payments/pull/439/commits/e58ab8f5bdf741461fde1cbcb3b889f573d149d6)

kimyou1102 committed:
refactor: InputHTMLAttributes타입에 정의되어 있는 중복된 필드 제거

kimyou1102 committed:
refactor: useId훅을 이용해서 고유한 id값이 되도록 수정

- CardValidityPeriodField 컴포넌트를 여러번 사용시 id가 중복되는 문제를 해결하기 위해 useId 사용

kimyou1102 committed:
feat: storybook autodocs 세팅 추가

kimyou1102 committed:
refactor: 스토리북에서 커스텀훅 적용

- Default 스토리에서 커스텀훅 사용해서 검증 및 사용자 입력 가능하도록 함
- Pass, Error 스토리에서는 입력 가능하지 않고 모양만 보이도록 함

kimyou1102 committed:
refactor: checkValideDate 함수 유틸함수로 분리

kimyou1102 committed:
fix: 유효기간 월 년 변경 시 잘못된 에러 표시가 남아 있던 문제를 해결

- 각 경우에서 isError를 false해줄 type을 넘겨줘서 해당 type이 isError가 되도록 수정

kimyou1102 committed:
fix: 현재 보다 이전날짜를 유효기간으로 입력했을 때 올바르지 않은 곳에 빨간표시되는 문제 해결

- 이전날짜 선택시 현재 입력한 input창에 빨간표시를 하게 해서 현재 입력창이 아닌 다른 곳이 문제일때 올바르지 않은 표시가 생기는 버그 발생
- 현재보다 이전인지 체크하는 함수에서 Boolean 반환 대신 문제된 필드(month 또는 year) 식별 로직으로 변경

kimyou1102 committed:
fix: 카드넘버 유효성 검사 이전에 입력한 인풋이 잘못됐을때 에러메시지 뜨지 않는 버그 수정

- 현재 value에 대한 검사만 했어서 다른 입력들에 대한 유효성 검사도 추가하여 변경

kimyou1102 committed:
refactor: label과 input연결

kimyou1102 committed:
refactor: 입력 길이 상수화

kimyou1102 commented on src/components/common/Input/Input.tsx:
```diff
@@ -0,0 +1,26 @@
+  onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
+  onBlur?: (e: React.FocusEvent<HTMLInputElement>) => void;
+  onFocus?: (e: React.FocusEvent<HTMLInputElement>) => void;
+}
```
수정커밋: [refactor: InputHTMLAttributes타입에 정의되어 있는 중복된 필드 제거](https://github.com/woowacourse/react-payments/pull/439/commits/39807ea35203413142d06e8814d9325ba679d7bf)

kimyou1102 commented:
안녕하세요 쵸파! 말씀해주신 리뷰들과 UX관련 기능들 수정했습니다. 덕분에 쵸파의 의견도 들어볼 수 있어서 좋았습니다!

## UX 수정 사항
- 유효기간 입력 검증 수정 커밋: [fix: 카드넘버 유효성 검사 이전에 입력한 인풋이 잘못됐을때 에러메시지 뜨지 않는 버그 수정](https://github.com/woowacourse/react-payments/pull/439/commits/9269e782f860894a515319dc3a8c2884e901ff1b)

## 추가 수정 사항
- 카드넘버 유효성 검사 버그 픽스: [fix: 카드넘버 유효성 검사 이전에 입력한 인풋이 잘못됐을때 에러메시지 뜨지 않는 버그 수정](https://github.com/woowacourse/react-payments/pull/439/commits/9269e782f860894a515319dc3a8c2884e901ff1b)
- 상수화 커밋: [refactor: 입력 길이 상수화](https://github.com/woowacourse/react-payments/pull/439/commits/289669f15f3f2b1c8082a94fe8750185391634f6)
## 추가 고민 사항
스토리북에 커스텀훅을 사용할 수 있냐는 쵸파의 피드백을 보고 스토리북을 수정했습니다.

처음엔 모든 스토리에 커스텀훅을 사용하니 초기값을 원하는 값으로 줄 수 없었습니다. 그래서 생각한 방식이 아래와 같습니다.

- Default: 아무것도 입력되지 않은 초기값. 커스텀훅을 사용하여 실제 컴포넌트와 유사하게 유효성검사가 일어나는 스토리
- Pass: 올바른 입력이 이루어진 스토리. args으로 초기값만 세팅했기에 입력이 변경되지 않음
- Error: 올바르지 않은 입력이 이루어진 스토리. args으로 초기값만 세팅했기에 입력이 변경되지 않음

"스토리북은 귀찮은 일을 덜어주기 위한 것이고 세부적인 검증 로직은 스토리북의 책임 범위가 아니다"라는 공식문서에 내용대로 Default를 제외한 스토리는 모양만 보여주게 했습니다. 
이런식으로 스토리를 작성해도 될지, 쵸파의 스토리북 작성 기준이 궁금합니다!

bassyu reviewed (approved):
안녕하세요 데이지!
답변이랑 반영 잘 고민해주셨네요~! 👍 

아직 카드 번호 입력에서는 에러 메세지가 나왔다가 다시 사라지는 부분이 있는데, 이 부분은 2단계에서 보완되면 좋겠습니다!

> 이런식으로 스토리를 작성해도 될지, 쵸파의 스토리북 작성 기준이 궁금합니다!

데이지 말대로 로직 테스트는 것은 Jest가 더 어울리겠네요!
스토리북은 UI의 다양한 상태나 여러 컴포넌트를 조합하면서 테스트와 문서화에 용이하다고 생각해요.

지금도 좋고, 보완 해보자면 Input의 더 다양한 상태나 props를 테스트 해볼 것 같아요. 
(예를 들어 `isDisabled`, `isError`, `isReadonly` props가 모두 true면 어떻게 보여야 할지?)
또는 [Input + InputFiled], [Input 여러개 + InputFiled + Label] 이런식으로 다양한 예시를 보여주면 좋을 것 같습니다~!

---

수고 많으셨습니다 데이지!
2단계도 파이팅입니다~! 💪 💪

woowapark commented:
@coderabbitai review  

##📌 역할: 
React 설계 원칙과 유지보수성을 기준으로, 시니어 프론트엔드 개발자 관점에서 설계 리뷰를 요청합니다.  
React 설계 원칙에 대한 깊은 이해와 풍부한 실무 경험을 함께 가지고 있는 시니어 웹 프론트엔드 개발자로 행동합니다.
당신의 임무는 이 PR을 작성한 주니어 개발자가 **React 설계 원칙을 이해하고**, **유지보수하기 좋은 코드**를 작성하는 실무적 역량을 키울 수 있도록 돕는 것입니다. 

**반드시 다음 가이드를 따라야 합니다**
• 명시적으로 요청하지 않는 한, 완전한 해결 코드를 제공하지 마세요.
• 정답을 바로 주는 대신, 시니어 개발자로서의 사고 과정을 드러내며 문제 해결 방법을 안내하세요.
• React 설계 원칙에 따라 리뷰할 때는, 왜 해당 원칙이 중요한지, 어떤 고민 과정을 거쳐야 하는지 이해할 수 있도록 도와주세요.
• 즉시 사용 가능한 해결책을 주기보다는, 관련 문서나 공식 레퍼런스를 참고하도록 안내하세요.

## 📝 요청 사항:  
다음 React 설계 원칙을 기준으로 코드가 잘 작성되었는지 평가해 주세요.  
1. 컴포넌트는 독립적이고 순수하게 작성되었는가? (Stateless, Pure Component를 지향하고 있는가?)  
필요 이상의 상태를 가지는 경우는 없는가?
컴포넌트의 동작이 명확한 입력(props)과 상태에 의해 결정되며, 불필요하거나 숨겨진 부작용(side effects)은 없는가?
   
2. Single Source of Truth(SSoT)에 따라 상태가 일관되게 관리되고 있는가?   
불필요한 로컬 상태를 지양하고, 상태의 소유 위치(lifting state up)를 고민해볼 부분은 없는가?

3. 단방향 데이터 흐름(one-way/unidirectional data flow)을 잘 유지하고 있는가?
상태 변경은 항상 상위에서 하위로, Props로 데이터를 전달하고, 콜백을 통해 상위로 알리는 패턴을 유지하고 있는가?
  
4. 선언적 UI 패턴으로 명령형 코드를 최소화했는가?  
React에서 추구하는 방향성에 맞추어 컴포넌트 UI를 더 선언적으로 구현하도록 고민해볼 부분은 없는가?

유지보수성과 확장성을 높이는 다음 특성을 고려해 주세요.  
- 재사용성 (Reusable)  
- 조합 용이성 (Composable)  
- 예측 가능성 & 디버깅 용이성 (Predictable & Easy to Debug)  
  - 복잡한 상태보다는 명확하고 단순한 상태 관리 패턴을 사용하고 있는가?
  - 에러 발생 시, 원인을 쉽게 추적할 수 있는 코드 흐름이 보장되는가?

🚨 이번 PR에서 반복적으로 나타난 React 설계 원칙 기준 Anti-Pattern 한 가지를 발견하고,  
왜 문제가 되는지, 어떻게 개선할 수 있는지 간략히 설명해 주세요:  
- Anti-Pattern:  
- 이유 & 개선 방향:  

## 출력 형식
결과는 한국어로, 아래 형식으로 요약해 주세요.

📚 [React 설계 원칙 준수 여부]  
- 독립성과 순수성:  
- SSoT 일관성:  
- 단방향 데이터 흐름:  
- 선언적 UI 구성:  

📋 [유지보수성 및 확장성]  
- 재사용성:  
- 조합 용이성:  
- 예측 가능성 & 디버깅 용이성:  

🚨 [반복적으로 나타난 React 설계 Anti-Pattern]

✨ [종합 평가 및 향후 개선 방향]

woowapark commented:
@coderabbitai review

📌 역할:
React 설계 원칙에 대한 깊은 이해와 풍부한 실무 경험을 함께 가지고 있는 시니어 웹 프론트엔드 개발자로 행동합니다.
당신의 임무는 이 PR을 작성한 주니어 개발자가 **React 설계 원칙을 이해하고**, **유지보수하기 좋은 코드**를 작성하는 실무적 역량을 키울 수 있도록 돕는 것입니다. 
아래 [1. 학습 목표 달성 여부], [2. 기능 요구 사항 충족 여부], [3. 프로그래밍 요구 사항 충족 여부] 세 가지 관점에서 상세하고 구체적인 리뷰를 요청합니다.

반드시 다음 가이드를 따라야 합니다:
•	명시적으로 요청하지 않는 한, 완전한 해결 코드를 제공하지 마세요.
•	정답을 바로 주는 대신, 시니어 개발자로서의 사고 과정을 드러내며 문제 해결 방법을 안내하세요.
•	React 설계 원칙에 따라 리뷰할 때는, 왜 해당 원칙이 중요한지, 어떤 고민 과정을 거쳐야 하는지 이해할 수 있도록 도와주세요.
•	즉시 사용 가능한 해결책을 주기보다는, 관련 문서나 공식 레퍼런스를 참고하도록 안내하세요.

---

[1. 학습 목표 달성 여부]  
(PR 본문 내 🎯 미션 소개 섹션 기준)

- [ ] 재사용 가능한 Input Component를 개발했는지?  
- [ ] Storybook을 사용하여 컴포넌트 상태 테스트가 잘 되어 있는지?  
- [ ] 카드 정보를 효과적으로 렌더링하기 위한 상태 관리가 적절하게 이루어졌는지?  

---

[2. 기능 요구 사항 충족 여부]  
- [ ] 카드 번호 입력 및 식별
  - [ ] 사용자가 입력하는 카드 번호를 실시간으로 파악하여, Visa나 MasterCard에 해당하면 해당 브랜드의 로고를 UI에 표시한다.
  - [ ] 입력은 숫자만 가능하며, 유효하지 않은 번호 입력 시 피드백을 제공한다.  
- [ ] 카드 유효기간 입력
  - [ ] 월과 년도를 범위 내에서만 입력할 수 있어야 하며, 입력 제한을 두어 사용자가 숫자만 입력할 수 있도록 한다.
- [ ] 실시간 프리뷰 업데이트
  - [ ] 사용자의 카드 정보 입력에 따라 카드 프리뷰를 동시에 업데이트한다.  

---

[3. 프로그래밍 요구 사항 충족 여부]  
(PR 본문 내 ✅ 리뷰어 체크 포인트 섹션 기준)

- [ ] **재사용성**과 **확장성**을 고려해 Input 컴포넌트를 잘 설계했는가?  
- [ ] 컴포넌트를 분리할 기준(역할, 책임, 관심사 등)을 고민한 흔적이 보이는가? 
- [ ] UI 단위가 지나치게 작거나 크지 않고 적절한 수준으로 구성되어 있는가? 
- [ ] 카드 정보 렌더링에 필요한 상태를 React의 상태로 적절하게 관리하고 있는가?
- [ ] 컴포넌트의 다양한 상태(정상 입력, 오류, 빈값 등)를 확인할 수 있는 스토리가 작성되어 있는가?
- [ ] 유효성 검증, 에러 메시지 등에서 사용자 경험을 고려한 처리가 이루어졌는가?  

---

📌 결과는 아래와 같은 형식으로 요약해 주세요:

📚 **[학습 목표 달성 여부]**  
- [x] 재사용 가능한 Input Component 개발:  
- [ ] Storybook 사용:  
- [x] 상태 관리 경험:  

🎯 **[기능 요구 사항 충족 여부]**  
{위와 유사한 체크 리스트}

📋 **[프로그래밍 요구 사항 충족 여부]**  
{위와 유사한 체크 리스트}

---

✨ 종합적으로 학습 목표, 기능 요구 사항, 프로그래밍 요구 사항 충족도를 %로 평가하고, 주요 개선 포인트를 정리해 주세요.

woowapark commented:
@coderabbitai review

📌 역할:
React 설계 원칙에 대한 깊은 이해와 풍부한 실무 경험을 함께 가지고 있는 시니어 웹 프론트엔드 개발자로 행동합니다.
당신의 임무는 이 PR을 작성한 주니어 개발자가 **React 설계 원칙을 이해하고**, **유지보수하기 좋은 코드**를 작성하는 실무적 역량을 키울 수 있도록 돕는 것입니다. 
아래 [1. 학습 목표 달성 여부], [2. 기능 요구 사항 충족 여부], [3. 프로그래밍 요구 사항 충족 여부] 세 가지 관점에서 상세하고 구체적인 리뷰를 요청합니다.

반드시 다음 가이드를 따라야 합니다:
• 명시적으로 요청하지 않는 한, 완전한 해결 코드를 제공하지 마세요.
• 정답을 바로 주는 대신, 시니어 개발자로서의 사고 과정을 드러내며 문제 해결 방법을 안내하세요.
• React 설계 원칙에 따라 리뷰할 때는, 왜 해당 원칙이 중요한지, 어떤 고민 과정을 거쳐야 하는지 이해할 수 있도록 도와주세요.
• 즉시 사용 가능한 해결책을 주기보다는, 관련 문서나 공식 레퍼런스를 참고하도록 안내하세요.

---

[1. 학습 목표 달성 여부]  
(PR 본문 내 🎯 미션 소개 섹션 기준)

- [ ] 재사용 가능한 Input Component를 개발했는지?  
- [ ] Storybook을 사용하여 컴포넌트 상태 테스트가 잘 되어 있는지?  
- [ ] 카드 정보를 효과적으로 렌더링하기 위한 상태 관리가 적절하게 이루어졌는지?  

---

[2. 기능 요구 사항 충족 여부]  
- [ ] 카드 번호 입력 및 식별
  - [ ] 사용자가 입력하는 카드 번호를 실시간으로 파악하여, Visa나 MasterCard에 해당하면 해당 브랜드의 로고를 UI에 표시한다.
  - [ ] 입력은 숫자만 가능하며, 유효하지 않은 번호 입력 시 피드백을 제공한다.  
- [ ] 카드 유효기간 입력
  - [ ] 월과 년도를 범위 내에서만 입력할 수 있어야 하며, 입력 제한을 두어 사용자가 숫자만 입력할 수 있도록 한다.
- [ ] 실시간 프리뷰 업데이트
  - [ ] 사용자의 카드 정보 입력에 따라 카드 프리뷰를 동시에 업데이트한다.  

---

[3. 프로그래밍 요구 사항 충족 여부]  
(PR 본문 내 ✅ 리뷰어 체크 포인트 섹션 기준)

- [ ] **재사용성**과 **확장성**을 고려해 Input 컴포넌트를 잘 설계했는가?  
- [ ] 컴포넌트를 분리할 기준(역할, 책임, 관심사 등)을 고민한 흔적이 보이는가? 
- [ ] UI 단위가 지나치게 작거나 크지 않고 적절한 수준으로 구성되어 있는가? 
- [ ] 카드 정보 렌더링에 필요한 상태를 React의 상태로 적절하게 관리하고 있는가?
- [ ] 컴포넌트의 다양한 상태(정상 입력, 오류, 빈값 등)를 확인할 수 있는 스토리가 작성되어 있는가?
- [ ] 유효성 검증, 에러 메시지 등에서 사용자 경험을 고려한 처리가 이루어졌는가?  

---

📌 결과는 아래와 같은 형식으로 요약해 주세요:

📚 **[학습 목표 달성 여부]**  
- [x] 재사용 가능한 Input Component 개발:  
- [ ] Storybook 사용:  
- [x] 상태 관리 경험:  

🎯 **[기능 요구 사항 충족 여부]**  
{위와 유사한 체크 리스트}

📋 **[프로그래밍 요구 사항 충족 여부]**  
{위와 유사한 체크 리스트}

✨ **[종합 평가]**

학습 목표 달성도: n%
{상세 항목}

기능 요구 사항 충족도: n%
{상세 항목}

프로그래밍 요구 사항 충족도: n%
{상세 항목}

주요 개선 포인트:

---

✨ 종합적으로 학습 목표, 기능 요구 사항, 프로그래밍 요구 사항 충족도를 %로 평가하고, 주요 개선 포인트를 정리해 주세요.

woowapark commented:
@coderabbitai review  

##📌 역할: 
React 설계 원칙과 유지보수성을 기준으로, 시니어 프론트엔드 개발자 관점에서 설계 리뷰를 요청합니다.  
React 설계 원칙에 대한 깊은 이해와 풍부한 실무 경험을 함께 가지고 있는 시니어 웹 프론트엔드 개발자로 행동합니다.
당신의 임무는 이 PR을 작성한 주니어 개발자가 **React 설계 원칙을 이해하고**, **유지보수하기 좋은 코드**를 작성하는 실무적 역량을 키울 수 있도록 돕는 것입니다. 

💡 **[학습 심화 유도]**  
- 현재 PR 코드에서, 아래 [생각해볼 질문들]과 연결하여 스스로 답을 찾아볼 수 있도록 유도 질문을 작성해 주세요.  
- **직접적인 답을 주지 말고**, 관련된 코드와 연관된 맥락에서 왜 이 질문이 중요한지 고민할 수 있도록 작성해 주세요.  
- 반드시 관련 코드 라인 번호를 함께 명시해 주세요.  
- 만약 현재 PR 코드에 직접 관련된 구현이 없다면,  
  - "어떤 상황에서 이런 구현이 필요했을까요?"  
  - "이번 과제에서 이러한 코드가 작성되지 않았을 때, 놓치고 있는 React 설계 포인트는 무엇일까요?"  
  와 같은 유도 질문을 작성해 주세요.

**[생각해볼 질문들]**  
1. 상태를 직접 변경하지 않고 굳이 `setState()`를 사용하는 이유가 무엇일까요?  
2. `setState()`는 비동기적으로 작동할까요?  
3. 불변성을 지켜야 하는 이유는 무엇일까요? 

결과는 질문별로 아래 형식으로 요약해 주세요.

🧩 [생각해볼 질문들과 관련된 코드 유도 질문]  
- [질문 3] 관련 코드: Line 110  
  - (예시) 이 상태 객체를 직접 변경하는 대신 새로운 객체를 만들어서 반환하는 방식으로 바꾼다면, 어떤 장점이 있을까요?  
- [질문 3] 관련 코드: ❌ 작성되지 않음  
  - (질문) 현재 코드에서 불변성을 고민해야 하는 부분이 없다면, 복잡한 상태 관리나 값 공유가 필요할 때 어떤 문제가 생길 수 있을지 예상해 보세요.
